### ISSUE TITLE
DateTimeOffset serialization does not preserve local date/time

### ISSUE THREAD

**Issue by russell-mcdaniel** [ch-robinson/dotnet-avro#29]
Because DateTimeOffset is converted to a UTC-based DateTime prior to serialization, DateTimeOffsets created from local times are not deserialized back to the same representation.

The unit tests do not catch this because the equality operator for DateTimeOffset returns true if the represented points in time are the same. It does not care *how* they are represented as [noted here](https://docs.microsoft.com/en-us/dotnet/api/system.datetimeoffset.op_equality).

To resolve this, DateTimeOffset should be handled separately from DateTime. The serializer and deserializer should use the ToString() and Parse() methods, respectively, on DateTimeOffset itself to preserve the original semantics of the DateTimeOffset value.

Reference:

BinarySerializerBuilder.cs / StringSerializerBuilderCase.BuildDelegate()
https://github.com/ch-robinson/dotnet-avro/blob/master/src/Chr.Avro.Binary/BinarySerializerBuilder.cs#L1953

BinaryDeserializerBuilder.cs / StringDeserializerBuilderCase.BuildDelegate()
https://github.com/ch-robinson/dotnet-avro/blob/development/src/Chr.Avro.Binary/BinaryDeserializerBuilder.cs#L1996

I'm still familiarizing myself with the code base here, but it appears that this implementation is consistent across the `master`, `development`, and `3.x` branches as of this writing. I was focused on the string serializer because strings are the recommended schema type for handling DateTimes and DateTimeOffsets in this library. I have not looked closely at the micro- and millisecond logical type implementations to see if / how they might be affected by this.

This code illustrates the problem:

```cs
var loc = DateTime.Now;
var utc = loc.ToUniversalTime();
var dtoFromLocal = new DateTimeOffset(loc);
var dtoFromUtc = new DateTimeOffset(utc);
```

Although .NET considers `dtoFromLocal` and `dtoFromUtc` to be "equal", note that the DateTime and Offset properties differ, so they are not *identical*, and this can affect how they are rendered and persisted by other applications.

This code demonstrates handling roundtrip the value identically:

```cs
var dtoIn = DateTimeOffset.Now;
var dtoStr = dtoIn.ToString("O");
var dtoOut = DateTimeOffset.Parse(dtoStr);
```

**Comment by dstelljes** [comment]
Thanks for the report; this is really helpful. The best course of action is probably to split apart the string serde builder cases (the string cases are currently too big, IMO) and, like you said, handle DateTimeOffset and DateTime separately.

It doesn’t appear that the timestamp logical type cases are affected by this—DateTime is converted to DateTimeOffset and the serializer builder uses DateTimeOffset.UtcTicks to calculate the difference with epoch. (Deserialized values are always UTC, which is why this library favors string representations.)

**Comment by dstelljes** [comment]
Fixed in #36 and released with 3.0.0.
